File: summary/go/24_segment_tree.go
package main

import (
	"fmt"
)

// セグメント木構造体
type SegmentTree struct {
	data []int // セグメント木のデータ
	n    int   // データの個数
}

// 新しいセグメント木を作成
func NewSegmentTree(n int) *SegmentTree {
	// データの個数が2のべき乗になるまで倍にする
	segtree := &SegmentTree{
		data: make([]int, 2*n-1),
		n:    n,
	}
	for i := n; i < 2*n-1; i++ {
		segtree.data[i] = 0 // 初期値を0に設定
	}
	return segtree
}

// 区間[l, r]の和を取得
func (segtree *SegmentTree) query(l, r, idx, a, b int) int {
	// 範囲外の場合は0を返す
	if b <= l || r <= a {
		return 0
	}

	// 区間が完全に含まれる場合は、データの値を返す
	if l <= a && b <= r {
		return segtree.data[idx]
	}

	// 部分的に含まれる場合は、子の値を再帰的に取得して合計する
	v1 := segtree.query(l, r, idx*2+1, a, (a+b)/2)
	v2 := segtree.query(l, r, idx*2+2, (a+b)/2, b)
	return v1 + v2
}

// 区間[l, r]の和を取得（main関数から呼び出す関数）
func Query(segtree *SegmentTree, l, r int) int {
	return segtree.query(l, r, 0, 0, segtree.n)
}

// 単一要素の更新
func (segtree *SegmentTree) update(i, val int) {
	// 更新対象のインデックスまで再帰的に値を更新
	idx := 0
	for i > 0 {
		segtree.data[idx] += val
		idx = idx*2 + (i & 1)
		i >>= 1
	}
	segtree.data[idx] += val
}

// main関数
func main() {
	// データの配列
	data := []int{1, 3, 2, 5, 4, 7}

	// セグメント木を作成
	n := len(data)
	segtree := NewSegmentTree(n)

	// データをセグメント木に挿入
	for i, v := range data {
		segtree.update(i, v)
	}

	// 区間[1, 3]の和を取得
	sum := Query(segtree, 1, 3)
	fmt.Println("区間[1, 3]の和:", sum) // 6 (1 + 3 + 2)

	// 要素2を10に更新
	segtree.update(2, 10)

	// 区間[0, 4]の和を取得
	sum = Query(segtree, 0, 4)
	fmt.Println("区間[0, 4]の和:", sum) // 22 (1 + 10 + 2 + 5 + 4)
}

File: summary/c++/24_segment_tree.cpp
#include <iostream>
#include <vector>

class SegmentTree {
private:
    std::vector<int> tree;
    int n;

    // Build the segment tree
    void build(const std::vector<int>& data, int node, int start, int end) {
        if (start == end) {
            tree[node] = data[start];
        } else {
            int mid = (start + end) / 2;
            build(data, 2 * node + 1, start, mid);
            build(data, 2 * node + 2, mid + 1, end);
            tree[node] = tree[2 * node + 1] + tree[2 * node + 2];
        }
    }

    // Update the segment tree
    void update(int node, int start, int end, int idx, int val) {
        if (start == end) {
            tree[node] = val;
        } else {
            int mid = (start + end) / 2;
            if (start <= idx && idx <= mid) {
                update(2 * node + 1, start, mid, idx, val);
            } else {
                update(2 * node + 2, mid + 1, end, idx, val);
            }
            tree[node] = tree[2 * node + 1] + tree[2 * node + 2];
        }
    }

    // Query the segment tree
    int query(int node, int start, int end, int L, int R) {
        if (R < start || end < L) {
            return 0;
        }
        if (L <= start && end <= R) {
            return tree[node];
        }
        int mid = (start + end) / 2;
        int leftSum = query(2 * node + 1, start, mid, L, R);
        int rightSum = query(2 * node + 2, mid + 1, end, L, R);
        return leftSum + rightSum;
    }

public:
    SegmentTree(const std::vector<int>& data) {
        n = data.size();
        tree.resize(4 * n);
        build(data, 0, 0, n - 1);
    }

    void update(int idx, int val) {
        update(0, 0, n - 1, idx, val);
    }

    int query(int L, int R) {
        return query(0, 0, n - 1, L, R);
    }
};

int main() {
    std::vector<int> data(5);
    data[0] = 1;
    data[1] = 2;
    data[2] = 3;
    data[3] = 4;
    data[4] = 5;

    SegmentTree segTree(data);

    // 区間 [1, 3] の和を求める
    std::cout << "Sum of range [1, 3]: " << segTree.query(1, 3) << std::endl;

    // 位置 2 の値を 10 に更新する
    segTree.update(2, 10);

    // 再び区間 [1, 3] の和を求める
    std::cout << "Sum of range [1, 3] after update: " << segTree.query(1, 3) << std::endl;

    return 0;
}
